### ISSUE TITLE
[incremental compiler] Support for adding classes.

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/fletch#72]
It seems the incremental compiler does not support adding new classes. We do have several incremental compiler tests which either make a program change which introduces a new class, or change code which makes of a class the first time:

```
# tests/fletch_tests/fletch_tests.status
incremental/encoded/add_class: Crash # VM Crash
incremental/encoded/newly_instantiated_class: Crash # VM Crash
incremental/encoded/newly_instantiated_class_X: Crash # VM Crash
incremental/encoded/newly_instantiated_class_with_fields: Crash # VM Crash
incremental/encoded/newly_instantiated_subclases_two_updates:Crash # VM Crash
incremental/encoded/newly_instantiated_superclasses_two_updates: Crash # VM Crash
incremental/encoded/super_classes_of_directly_instantiated: Crash # VM Crash
```

I've spent some time debugging the first `add_class` VM crash, and it seem the other tests mentioned are hitting the same problem.

The `add_class` test introduces a new class in a code update and references the constructor of the new class in `main`. The compiler pushes the following incremental commands to the VM:
```
CommandCode.PrepareForChanges()
CommandCode.PushNewFunction(1, 0, [load literal null, return 1 1, method end 4], [])
CommandCode.PopToMap(MapId.methods, 377)
CommandCode.PushNull()
CommandCode.PushNull()
CommandCode.PushNewFunction(1, 2, [load const @0, invoke static @1, pop, load literal null, return 1 1, method end 15], [])
CommandCode.PopToMap(MapId.methods, 376)
CommandCode.PushNull()
CommandCode.PushNewFunction(0, 1, [invoke static @0, invoke 217088, pop, load literal null, return 1 0, method end 15], [])
CommandCode.PopToMap(MapId.methods, 10)
CommandCode.PushNewString('v2')
CommandCode.PopToMap(MapId.constants, 72)
CommandCode.PushFromMap(MapId.methods, 376)
CommandCode.PushFromMap(MapId.constants, 72)
CommandCode.ChangeMethodLiteral(0)
CommandCode.PushFromMap(MapId.methods, 376)
CommandCode.PushFromMap(MapId.methods, 11)
CommandCode.ChangeMethodLiteral(1)
CommandCode.PushFromMap(MapId.methods, 10)
CommandCode.PushFromMap(MapId.methods, 375)
CommandCode.ChangeMethodLiteral(0)
CommandCode.PushFromMap(MapId.methods, 19)
CommandCode.PushFromMap(MapId.methods, 10)
CommandCode.ChangeMethodLiteral(0)
```

This causes the fletch-vm to hit an assert:
```
../../src/vm/object.h:1673: error: expected: object->IsFunction()
Current stacktrace:
out/DevelopX64/fletch-vm(fletch::DynamicAssertionHelper::DumpStacktrace()+0x30) [0x4a12d8]
out/DevelopX64/fletch-vm(fletch::DynamicAssertionHelper::Fail(char const*, ...)+0x120) [0x4a1258]
out/DevelopX64/fletch-vm(fletch::Function::cast(fletch::Object*)+0x56) [0x4644fc]
out/DevelopX64/fletch-vm(fletch::Program::static_method_at(int) const+0x2c) [0x466708]
out/DevelopX64/fletch-vm(fletch::Engine::Interpret(fletch::TargetYieldResult*)+0x1d32) [0x468f3e]
out/DevelopX64/fletch-vm(fletch::Interpreter::Run()+0x106) [0x46c964]
out/DevelopX64/fletch-vm(fletch::Scheduler::InterpretProcess(fletch::Process*, fletch::ThreadState*)+0x76) [0x48f5ee]
out/DevelopX64/fletch-vm(fletch::Scheduler::RunInThread()+0x313) [0x48f431]
out/DevelopX64/fletch-vm(fletch::Scheduler::RunThread(void*)+0x20) [0x4903e8]
out/DevelopX64/fletch-vm(fletch::ThreadPool::RunThread(void*)+0x29) [0x4a0535]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x8182) [0x7f157d508182]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f157c81147d]
```

The reason we hit this assert is that in the following code is setting literal `0` of the `main` function to the `new C()` constructor, which doesn't seem to exist:
```
// main.literal[0] = "new C()"
CommandCode.PushFromMap(MapId.methods, 10)
CommandCode.PushFromMap(MapId.methods, 375)
CommandCode.ChangeMethodLiteral(0)
```

The `PushFromMap` implementation does **not validate**, that the map entry actually exists as can be seen in the following code snippet. The default `NULL` value for `none` gets returned:

```c++
 600 void Session::PushFromMap(int index, int64 id) {
 601   Push(maps_[index]->LookupById(id));
 602 }
---------------
 25   Object* LookupById(int64 id, Object* none = NULL);
---------------
 71 Object* ObjectMap::LookupById(int64 id, Object* none) {
 72   int index = BucketIndexFromId(id);
 73   Bucket* current = table_by_id_[index];
 74   while (current != NULL) {
 75     if (current->id == id) return current->object;
 76     current = current->next;
 77   }
 78   return none;
 79 }
```

This causes a `NULL` pointer to be inserted into `Array* Program::static_methods_` which causes us to hit the ASSERT (`0` is a Smi and not a `Function*`).

Looking through the incremental compiler revealed a TODO to implement the functionality for adding classes:
```dart
// pkg/fletchc/lib/incremental/library_updater.dart
 860     // TODO(ahe): Compute this.
 861     Set<ClassElementX> newClasses = new Set();
 862 
```

To summarize:
  - **fletch-vm** session handling code should be changed to validate that `PushFromMap` commands are referencing *existing* entries in a `ObjectMap`.
  - **incremental compiler** should be fixed to properly handle new classes in code updates.

**Comment by mkustermann** [comment]
@peter-ahe-google Can we close this issue?
